### ISSUE TITLE
Layout Component Definitions

### ISSUE THREAD

**Issue by FranciscoGileno** [inloco/supernova-inloco#35]
This intends to open a discussion about the Layout Component.

The objective is to define together with UI/UX team the scope and interface of this component.

Inloco Ads already implemented its layout, but we will need to refactor/improve it to be reusable across apps.

![image](https://user-images.githubusercontent.com/1139664/47604559-e220e900-d9d1-11e8-9b72-47015443a219.png)

My initial interface proposal is:

```jsx
<Layout>
  <Layout.Header logo={logo}>
     {RightSideComponent}
  </Layout.Header>
  <Layout.Sidebar color={color}>
    <Layout.SidebarItem 
       tooltip={tooltip} 
       selected={selected} 
       icon={icon} 
       href={href} 
    />
  </Layout.Sidebar>
  <Layout.Main>
    <Layout.Title title={title} breadcrumb={breadcrumb}>
      {RightSideComponent}
    </Layout.Title>
    {MainChildren}
  </Layout.Main>
</Layout>
```

Of course, all components will bypass extra props to its internals parent, so for example, we could add the prop `className` to any of them.

There are a few questions yet I don't know the answer:

- How the layout works on mobile / tablet screens?
- Can a Sidebar item be disabled? 
- What can change from project to project, beside colors?
- When scrolling, the header always stays visible, in any resolution?

Please, let's discuss. 

Should we add/remove responsability from this component? 
Should we change the contract of any of these components?

@mairatma @peaonunes @TCezarRod @kaueburiti @bruno-perruci

**Comment by TCezarRod** [comment]
This is already my favourite component. 😄 

About the colours, I don't think it would change between projects. I believe the only reason this is happening today is because the older projects are still using the classic supernova which still has the old colors, but this should change with this new shared Layout. The colours should be the same shades of Neutral in all projects, is that right, @bruno-perruci?

Some suggestions:
## Sidebar
* The Sidebar could also have Sections. Two of our products use sections to divide the contents on the side bar. Maybe something like this could work, what do you think?
```
<Layout.Sidebar >
    <Layout.SidebarSection ... >
        <Layout.SidebarItem ... />
    </Layout.SidebarSection>
    <Layout.SidebarSection ... >
        <Layout.SidebarItem ... />
        <Layout.SidebarItem ... />
    </Layout.SidebarSection>    
  </Layout.Sidebar>
```

This would render that horizontal separator between sections.
<p align="center"> 
<img src="https://user-images.githubusercontent.com/9112403/47654180-0a8f1b80-db69-11e8-8390-f8160cde8781.png">
</p>

* It also seems a good idea to include a `Footer` on the sidebar as well. Engage uses it, and we have plans to include a similar Sidebar Item on places as well.

## Title
* The classic supernova keeps both header and title visible when scrolling, I vote for keeping that behaviour.

* What are your thoughts on that `breadcrumb` prop? I don't know if the other projects have this, but in Places we have a page with a "progress bar" where the breadcrumbs would be, so maybe that prop could be something more flexible and allow us to add any custom component to be under the Title, instead of calling it 'breadcrumbs' exclusively. What do you think?

## Screen sizes
* I think it is nice to think about responsiveness to screen sizes. Some components on the classic supernova did implement that in some components (including Layout) and I think it is a nice feature not to lose. But are the tools being built having multiple screen sizes in mind? If only the Layout is working for smaller screens but the remaining parts aren't, maybe we could leave that out for the first version of the Layout and add these changes later.

**Comment by FranciscoGileno** [comment]
Awesome comments @TCezarRod .

About sidebar, I am not sure if adding another wrapper just to divide content is a good idea. What if we have a `<Layout.Divider />` ?

I agree about the Title. `breadcrumb` might be to specific. We have something like that:
![image](https://user-images.githubusercontent.com/1139664/47661583-0fa79700-db78-11e8-8def-a6b5069fb135.png)

I could think a better way to add this customization.

**Comment by TCezarRod** [comment]
Oh, the position is different from what we have on Places. Our "secondary" component stays below the Title, like that:
![captura de tela 2018-10-29 as 13 02 46](https://user-images.githubusercontent.com/9112403/47663090-f6ecb080-db7a-11e8-9e5c-d50a91184e00.png)

We could keep this shorthand option as proposed, but also allow us to defined the content via `children`, just like some components on `react-semantic-ui` do. Something like that:
```
<Layout.Header title={title} breadcrumb={breadcrumb}>
    {RightSideComponent}
</Layout.Header>
```
or 
```
<Layout.Header>
    <Layout.Title breadcrumb={breadcrumb}>
        {Title}
    </Layout.Title>
    {RightSideComponent}
</Layout.Header>
```

These options would render the same thing, but on the latter one we could also add custom content inside `<Layout.Title>` (In our case it would be the progress bar). Or maybe a better structure, to avoid rendering components inside something called `Title` that should be purely text. What do you think?

We could use the name `Layout.Header`, and rename the proposed `Layout.Header` to `Layout.Navbar`, that's how we call them on the classic supernova, and I think it makes more sense. What do you think?

---

The `<Layout.Divider />` is really a better idea! We really don't have any behaviour that would be shared among the items in the section, so there's no need to group them. The Divider would work just fine, and keep everything cleaner 😄

**Comment by bruno-perruci** [comment]
> This is already my favourite component. 😄
> 
> About the colours, I don't think it would change between projects. I believe the only reason this is happening today is because the older projects are still using the classic supernova which still has the old colors, but this should change with this new shared Layout. The colours should be the same shades of Neutral in all projects, is that right, @bruno-perruci?
> 
> Some suggestions:
> 
> ## Sidebar
> * The Sidebar could also have Sections. Two of our products use sections to divide the contents on the side bar. Maybe something like this could work, what do you think?
> 
> ```
> <Layout.Sidebar >
>     <Layout.SidebarSection ... >
>         <Layout.SidebarItem ... />
>     </Layout.SidebarSection>
>     <Layout.SidebarSection ... >
>         <Layout.SidebarItem ... />
>         <Layout.SidebarItem ... />
>     </Layout.SidebarSection>    
>   </Layout.Sidebar>
> ```
> This would render that horizontal separator between sections.
> 
> ![](https://user-images.githubusercontent.com/9112403/47654180-0a8f1b80-db69-11e8-8390-f8160cde8781.png)
> 
> * It also seems a good idea to include a `Footer` on the sidebar as well. Engage uses it, and we have plans to include a similar Sidebar Item on places as well.
> 
> ## Title
> * The classic supernova keeps both header and title visible when scrolling, I vote for keeping that behaviour.
> * What are your thoughts on that `breadcrumb` prop? I don't know if the other projects have this, but in Places we have a page with a "progress bar" where the breadcrumbs would be, so maybe that prop could be something more flexible and allow us to add any custom component to be under the Title, instead of calling it 'breadcrumbs' exclusively. What do you think?
> 
> ## Screen sizes
> * I think it is nice to think about responsiveness to screen sizes. Some components on the classic supernova did implement that in some components (including Layout) and I think it is a nice feature not to lose. But are the tools being built having multiple screen sizes in mind? If only the Layout is working for smaller screens but the remaining parts aren't, maybe we could leave that out for the first version of the Layout and add these changes later.

**Comment by bruno-perruci** [comment]
> The colours should be the same shades of Neutral in all projects, is that right, @bruno-perruci?

As all of our projects use the same color for the menu I don't think we need to let it custom. Also, if we decide to change its color it'll be made in consense with all projects too.

> Should we add/remove responsability from this component?

Probably the first project that'll be responsive is the Ads, and I don't know when it'll be the main priority so IMO we'll need it in the future, but not yet. But, if you tell me that is better to do it responsive now I can create the design and share with all of you.


One thing that I think is missing on this component today is a submenu. We'll need to use it on Engage and I don't know if we already have it developed. Something like that:
![image](https://user-images.githubusercontent.com/43831090/47818892-b0908080-dd37-11e8-9130-3b8413b160e9.png)

This is an old print that needs some changes yet. Btw, I'll move forward and create an InVision project with this interaction and, if you tell me that is better to create the responsiveness from the beginning I add it there too.

Let me know what you think about @FranciscoGileno @TCezarRod @mairatma @kaueburiti @peaonunes

**Comment by mairatma** [comment]
Love it love it love it! ❤️ 

@bruno-perruci Having all that on invision would be great! By the way, let me know when you add the submenu to invision too. I tried to implement this with the old supernova in Engage, but didn't know some details, such as how should the `selected` and `hover` states look like. It's already available and working there, but could probably look better.

**Comment by FranciscoGileno** [comment]
@bruno-perruci It would be nice to have invision design of the mobile version.

Even if we do not code this now, knowing how it will be on mobile, we will consider it when writing the code.

It would be nice if you could detail the Title as well... What are the possibilities? normal, with breadcrumb, with progress bar, etc... This way we can include those behaviors.

Thanks!

**Comment by mairatma** [comment]
I'm officially working on this now guys! I'll take all the feedback gathered here in mind for the component's structure, and we can discuss more on the prs when I send them 😄 

I'll try to make things as flexible as possible according to what could be different between different projects (Bruno has answered some questions I made about this, we can ask more when necessary) but still have it be simple to use.

I think responsiveness is going to be for later though. We've decided to only support monitors from 1280px on for now.
